content/blog/2022/2022-02-14-spring-shell.md

```diff
@@ -0,0 +1,295 @@
+```java
+@SpringBootApplication  
+public class SpringShellApplication {  
+   public static void main(String[] args) {  
+      SpringApplication.run(SpringShellApplication.class, args);  
+   }  
+}
+```
+
+Now, let's create our first shell command which simulates an SSH:
+```java
+@ShellComponent  
+public class SSHCommand {  
+    Logger log = Logger.getLogger(SSHCommand.class.getName());  
+  
+    @ShellMethod(value = "connect to remote server")  
+    public void ssh(@ShellOption(value = "-s") String remoteServer)  
+    {  
+        log.info(format("Logged to machine '%s'", remoteServer));  
+    }  
+}
+```
+
+The annotation `@ShellComponent` as defined by the Spring team:
+
+> Indicates that an annotated class may contain shell methods (themselves annotated with ShellMethod) that is, methods that may be invoked reflectively by the shell
+
+As for the `@ShellMethod` annotation:
+
+    Used to mark a method as invokable via Spring Shell.
+We can also see the `value` property which is used to describe the command.
+
+The `@ShellOption` annotation simply states that this command takes a parameter named `-s`.
+
+So as a result when we run the application we get a shell that has a command called `ssh` which takes a parameter `-s` and all it does is logging the passed parameter value to the command line.
+
+```bash
+shell:>ssh -s my-machine
+2022-02-11 15:44:04.065  INFO 5648 --- [           main] j.t.springshell.command.SSHCommand       : Logged to machine 'my-machine'
+shell:>
+```
+
+## Deeper Look Into Spring Shell
+### Command Name
+The default naming convention for Spring Shell, as we've seen, is taking the method name `ssh` and turning it into the command name.
+- If we wrote the name in camel case Spring would turn camelCase humps into "-".
+- So `customSsh` would translate to `custom-ssh`.
+
+We can also add a name of our own using the `key` property of the `ShellMethod` annotation:
+
+```java 
+@ShellMethod(key = "my-ssh", value = "connect to remote server")  
+public void ssh(@ShellOption(value = "-s") String remoteServer)  
+{  
+    log.info(format("Logged to machine '%s'", remoteServer));  
+}
+```
+
+### Command Parameters
+
+#### Parameter Naming
+As we've seen from the previous example, command parameters are expressed through method parameters.
+
+We can specify the name of the parameter using the `value` property of the `@ShellOption` annotation.
+
+If we don't specify the value however, Spring Shell assigns it a default value of parameter name "-" separated prefixed by `ShellMethod.prefix()`.
+- The `ShellMethod.prefix()` default value is "--" unless changed in the `@ShellMethod` annotation:
+```java
+@ShellMethod(key = "my-ssh", prefix = "-", value = "connect to remote server")  
+public void ssh(@ShellOption String remoteServer)  
+{  
+    log.info(format("Logged to machine '%s'", remoteServer));  
+}
+```
+- In the above example the parameter name will be `-remote-server`.
+
```

Comment:
Maybe show an example of what the parameter would look like in the shell? Like you did in the other sections?